### PR TITLE
Issue 126 turn key training demo

### PR DESCRIPTION
**Pull request by cdicle**
This PR implements a turn-key training demo-script using hydra config management. The goal is to reduce the entry barrier for the developer. Hydra enables config composition and convenient overwrite from command line which are demonstrated below. It is a widely used config management tool in ML community. The proposed workflow for the developer looks as below.

The developer 
1. clones the repo
2. creates the environment
3. downloads the input.wav and output.wav from provided links. (This step can be automated as well)
4. run demo by calling below from the repo root
```
python  nam/train/demo.py --config-path=bin/train/inputs/data --config-name=demo.yaml common.x_path=path/to/input.wav common.y_path=path/to/output.wav common.train_path=path/to/training_output"
```

Note that, hydra allows conveniently overwriting config parameters.

### CODE REVIEW

**Comment by sdatkinson** [comment]
Hi! Thanks for the PR.

What's the difference between this and using [`bin/train/main.py`](https://github.com/sdatkinson/neural-amp-modeler/blob/main/bin/train/main.py)?

* `demo.yaml` seems to copy [`bin/train/inputs/data/single_pair.json`](https://github.com/sdatkinson/neural-amp-modeler/blob/main/bin/train/inputs/data/single_pair.json)
* This doesn't seem to expose for the demo how the model is configured (cf e.g. [`bin/train/inputs/models/wavenet.json`](https://github.com/sdatkinson/neural-amp-modeler/blob/main/bin/train/inputs/models/wavenet.json) or other model configs in there) nor the learning algorithm.

Do you think that we might achieve effectively the same result as this PR by sprucing up the README? (I also wouldn't mind not picking up another dependency with Hydra. What's the sell there?)

I think that, at a high level, I'm not convinced that this repo should be going all-in on the wrappers in `nam.train`, though that's obviously what `easy_colab.ipynb` and the GUI trainer do...basically if a developer wants to get into this, having something that doesn't conflate the UX with the moving parts underneath may be more helpful?

But obviously I have a PoV bias, so I'm interested to hear if you've got any feedback on these thoughts.

**Comment by cdicle** [comment]
@sdatkinson You are right this PR achieves the same functionality as [bin/train/main.py](https://github.com/sdatkinson/neural-amp-modeler/blob/main/bin/train/main.py). I have missed that. There are different entry points to the training, from colab, from command-line ...etc. I was pursuing the colab's call path. In that sense, yes, we can achieve similar affect if we improve the README, plus improve config management (more details below). I would also suggest populating all training functionality into single python file and calling relevant functions from the same colab, command line or gui. The idea is reducing the call pathways as much as possible to improve maintainability. 

About config management and hydra. In this example I only illustrate the command-line override convenience, where one can override a config parameter without need to change the config file. In this particular example, user can run the training without touching the config file. More powerful part of Hydra is composing configs. It allows you define default behavior and change details as needed. For example you can define all training parameters like optimizer, learning rate, batch size ...etc. in a config file but may prefer to change only the batch size for a particular model. Using hydra you can create a config file and override only the parts of the default config, the batch size in this case. It allows to switch between configs conveniently as well. The idea here is again reducing config duplication, or sources of truth as much as possible to improve maintainability. See [here](https://hydra.cc/docs/intro/#composition-example) for a composition example. Hydra works with yaml files that is why I converted `single_pair.json` to `demo.yaml`.

I hope this clarifies some of the questions above.

**Comment by cdicle** [comment]
Closing this in the light of our discussion and follow up [PR](https://github.com/sdatkinson/neural-amp-modeler/pull/150).
